### DISCUSSION TITLE
Unit test for abstract classes in projects generated by the CLI

### DISCUSSION THREAD

**Discussion by meysammahfouzi** [FilledStacks/stacked#1016] [Q&A]
First I created a project:

```dart
stacked create app bluesky
```

Then added a view:

```dart
stacked create view feed
```

Then added a service:

```dart
stacked create service http_api
```

After that I modified `FeedViewModel` to have a reference to the service:

```dart
class FeedViewModel extends BaseViewModel {
  HttpApiService httpApiService = locator<HttpApiService>();
}
```

Then wrote a unit test for the viewmodel and it passed ✅  (`feed_viewmodel_test.dart`):

```dart
import '../helpers/test_helpers.dart';

FeedViewModel _getModel() => FeedViewModel();

void main() {
  group('FeedViewModel Tests -', () {
    setUp(() => registerServices());
    tearDown(() => locator.reset());
    test('test', () async {
      final model = _getModel();
    });
  });
}
```

Now I decided to abstract the api service. So I manually created the following file: `api_service.dart`:

```dart
abstract class ApiService{
  Future<bool> likePost(int postId);
}
```

And changed the `HttpApiService` class:

```dart
class HttpApiService implements ApiService {
  @override
  Future<bool> likePost(int postId) {
    // TODO: implement likePost
    throw UnimplementedError();
  }
}
```

And in the `app.dart` file, changed the following line:

```dart
LazySingleton(classType: HttpApiService),
```

To the following:

```dart
LazySingleton(classType: HttpApiService, asType: ApiService),
```

And also updated the `FeedViewModel` class to locate the `ApiService` instead of `HttpApiService`:

```dart
class FeedViewModel extends BaseViewModel {
  ApiService apiService = locator<ApiService>();
}
```

And also ran the `stacked generate` command to make sure all of the files are correct.

Then I ran the unit test again and it failed 🔴  with the following error:

```bash
Bad state: GetIt: Object/factory with type ApiService is not registered inside GetIt. 
```

To fix this, I modified the `test_helpers.dart` file and executed `stacked generate`:

```dart
MockHttpApiService getAndRegisterHttpApiService() {
  _removeRegistrationIfExists<HttpApiService>();
  final service = MockHttpApiService();
  locator.registerSingleton<ApiService>(service);
  return service;
}
```

And the test passed ✅ 

Did I modify the correct files? Is there something I should have done differently?

**Comment by ferrarafer** [comment]
Hi @meysammahfouzi should be like below.

```dart
MockApiService getAndRegisterApiService() {
  _removeRegistrationIfExists<ApiService>();
  final service = MockApiService();
  locator.registerSingleton<ApiService>(service);
  return service;
}
```

Locator will do the rest under the hood.
